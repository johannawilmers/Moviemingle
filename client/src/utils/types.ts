/**
 * Interface representing a movie.
 * @interface
 * @property {number | undefined} _id - The unique identifier of the movie.
 * @property {string} title - The title of the movie.
 * @property {string} original_title - The original title of the movie.
 * @property {string} poster_path - The path to the movie's poster image.
 * @property {string} backdrop_path - The path to the movie's backdrop image.
 * @property {number[]} genre_ids - An array of genre IDs associated with the movie.
 * @property {string} overview - A brief overview or description of the movie.
 * @property {string} release_date - The release date of the movie.
 * @property {number} vote_average - The average vote or rating for the movie.
 * @property {boolean | undefined} isFavorite - Whether the movie is marked as a favorite.
 */
export interface movie {
  _id: number | undefined;
  title: string;
  original_title: string;
  poster_path: string;
  backdrop_path: string;
  genre_ids: number[];
  overview: string;
  release_date: string;
  vote_average: number;
  isFavorite?: boolean;
}

/**
 * Interface representing a genre.
 * @interface
 * @property {number | undefined} _id - The unique identifier of the genre.
 * @property {string} name - The name of the genre.
 */
export interface genre {
  _id: number | undefined;
  name: string;
}
