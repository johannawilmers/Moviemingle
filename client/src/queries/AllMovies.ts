import { gql } from '@apollo/client';

const GET_MOVIES = gql`
  query getMovies($skip: Int, $limit: Int) {
    getMovies(skip: $skip, limit: $limit) {
      _id
      title
      original_title
      poster_path
      genre_ids
      overview
      release_date
      vote_average
      backdrop_path
    }
  }
`;

const GET_SINGLE_MOVIE = gql`
  query getMovie($id: Int!, $userID: String!) {
    getMovie(ID: $id, userID: $userID) {
      _id
      title
      original_title
      poster_path
      genre_ids
      overview
      release_date
      vote_average
      backdrop_path
      isFavorite
    }
  }
`;

const GET_FILTERED_MOVIE_COUNT = gql`
  query getFilteredMoviesCount($genre_id: Int, $search: String, $sortValue: String) {
    getFilteredMoviesCount(genre_id: $genre_id, search: $search, sortValue: $sortValue) {
      total
    }
  }
`;

function getFilteredMovies() {
  const GET_FILTERED_MOVIES = gql`
    query getFilteredMovies(
      $genre_id: Int
      $skip: Int
      $limit: Int
      $search: String
      $sortValue: String
      $userID: String
    ) {
      getFilteredMovies(
        genre_id: $genre_id
        skip: $skip
        limit: $limit
        search: $search
        sortValue: $sortValue
        userID: $userID
      ) {
        _id
        title
        original_title
        poster_path
        genre_ids
        overview
        release_date
        vote_average
        backdrop_path
        isFavorite
      }
    }
  `;
  return GET_FILTERED_MOVIES;
}

export { GET_FILTERED_MOVIE_COUNT, GET_MOVIES, GET_SINGLE_MOVIE, getFilteredMovies };
