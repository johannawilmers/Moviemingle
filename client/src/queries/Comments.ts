import { gql } from '@apollo/client';

const ADD_COMMENT = gql`
  mutation AddComment($addCommentId: Int!, $comment: String!, $userID: String) {
    addComment(id: $addCommentId, comment: $comment, userID: $userID) {
      comment
      userID
    }
  }
`;

const GET_ALL_COMMENTS = gql`
  query GetAllComments($movieId: Int!) {
    getAllComments(movieID: $movieId) {
      comment
      userID
    }
  }
`;

export { ADD_COMMENT, GET_ALL_COMMENTS };
