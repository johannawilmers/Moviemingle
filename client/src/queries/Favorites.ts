import { gql } from '@apollo/client';

function getFavoritesByUserID() {
  const GET_FAVORITES_BY_USER_ID = gql`
    query GetFavoritesByUserID($userID: String!) {
      getFavoritesByUserID(userID: $userID) {
        _id
        title
        original_title
        poster_path
        genre_ids
        overview
        release_date
        vote_average
        backdrop_path
        isFavorite
      }
    }
  `;
  return GET_FAVORITES_BY_USER_ID;
}

const ADD_FAVORITE = gql`
  mutation AddFavorite($userID: String!, $id: Int!) {
    addFavorite(userID: $userID, id: $id) {
      _id
      userID
    }
  }
`;

const REMOVE_FAVORITE = gql`
  mutation RemoveFavorite($userID: String!, $id: Int!) {
    removeFavorite(userID: $userID, id: $id) {
      _id
      userID
    }
  }
`;

const IS_FAVORITE = gql`
  query IsFavorite($userID: String!, $id: Int!) {
    isFavorite(userID: $userID, id: $id)
  }
`;
export { ADD_FAVORITE, getFavoritesByUserID, IS_FAVORITE, REMOVE_FAVORITE };
