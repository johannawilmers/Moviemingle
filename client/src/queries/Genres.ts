import { gql } from '@apollo/client';

const GET_GENRE = gql`
  query getGenre($id: ID!) {
    getGenre(id: $id) {
      _id
      name
    }
  }
`;

const GET_ALL_GENRES = gql`
  query getAllGenres {
    getAllGenres {
      _id
      name
    }
  }
`;
export { GET_ALL_GENRES, GET_GENRE };
