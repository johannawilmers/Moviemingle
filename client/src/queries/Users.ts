import { gql } from '@apollo/client';

const ADD_USER = gql`
  mutation AddUser($userID: String!) {
    addUser(userID: $userID) {
      userID
      favorites {
        _id
        title
        original_title
        poster_path
        genre_ids
        overview
        release_date
        vote_average
        backdrop_path
      }
    }
  }
`;

const GET_ALL_USERS = gql`
  query GetAllUsers {
    getAllUsers
  }
`;

export { ADD_USER, GET_ALL_USERS };
