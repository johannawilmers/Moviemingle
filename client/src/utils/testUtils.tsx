import { ApolloClient, ApolloProvider, HttpLink, InMemoryCache } from '@apollo/client';
import { MockedProvider } from '@apollo/client/testing';
import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';

/**
 * HTTP link for Apollo Client to connect to a GraphQL server.
 */
const link = new HttpLink({
  uri: 'http://localhost:4000',
});

/**
 * Apollo Client instance configured with an in-memory cache and HTTP link.
 * @type {import('@apollo/client').ApolloClient<import('@apollo/client').NormalizedCacheObject>}
 */
const client = new ApolloClient({
  cache: new InMemoryCache(),
  link: link,
});

/**
 * Renders a component with Apollo Client and BrowserRouter for testing page content.
 * @function
 * @param {JSX.Element} component - React component to render.
 * @returns {RenderResult} - Result of rendering with testing utilities.
 */
export const renderWithPageContent = (component: JSX.Element) =>
  render(
    <ApolloProvider client={client}>
      <BrowserRouter>{component}</BrowserRouter>
    </ApolloProvider>,
  );

/**
 * Renders a component with Apollo MockedProvider for testing with mocked GraphQL data.
 * @function
 * @param {React.ReactNode} child - React component to render within the provider.
 * @param {any[]} mocks - Array of GraphQL query/mutation mocks.
 * @returns {RenderResult} - Result of rendering with testing utilities.
 */
export const renderWithMocks = (child: React.ReactNode, mocks: any[] = []) => {
  return render(
    <MockedProvider mocks={mocks} addTypename={false}>
      <BrowserRouter>{child}</BrowserRouter>
    </MockedProvider>,
  );
};
