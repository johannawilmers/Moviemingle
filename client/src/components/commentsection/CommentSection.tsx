import { useMutation, useQuery } from '@apollo/client';
import { Avatar, Button, Grid, Paper, TextField } from '@mui/material';
import Alert from '@mui/material/Alert';
import Box from '@mui/material/Box';
import CircularProgress from '@mui/material/CircularProgress';
import { blue } from '@mui/material/colors';
import { useState } from 'react';

import { ADD_COMMENT, GET_ALL_COMMENTS } from '../../queries/Comments';

/**
 * CommentSectionProps for the CommentSection component.
 */
interface CommentSectionProps {
  movieId: number;
  userID: string;
}

/**
 * CommentSection component for handling comments on a movie.
 *
 * @param {CommentSectionProps} props - The CommentSectionProps for the CommentSection component.
 * @returns {JSX.Element} The rendered CommentSection component.
 */
const CommentSection: React.FC<CommentSectionProps> = ({ movieId, userID }) => {
  const [newComment, setNewComment] = useState<string>('');
  const [addComment, { loading, error }] = useMutation(ADD_COMMENT);
  const { data: allCommentsData, refetch } = useQuery(GET_ALL_COMMENTS, {
    variables: { movieId: movieId },
  });

  const allComments = allCommentsData?.getAllComments || [];
  const [showAllComments, setShowAllComments] = useState(false);

  /**
   * Toggles the visibility of all comments.
   */
  const handleToggleComments = () => {
    setShowAllComments(!showAllComments);
  };

  /**
   * Updates comments on submission.
   *
   * @param {string} newComment - The new comment to be added.
   */
  const updateComments = async (newComment: string) => {
    if (newComment === '') {
      alert('You must write a comment.');
      return;
    }
    await addComment({
      variables: { addCommentId: movieId, comment: newComment, userID: userID },
      refetchQueries: [{ query: GET_ALL_COMMENTS, variables: { movieId: movieId } }],
    });

    refetch();
    setNewComment('');
  };

  /**
   * Handles the Enter key press event to submit comments.
   *
   * @param {React.KeyboardEvent} event - The keyboard event.
   */
  const handleEnterPress = (event: React.KeyboardEvent) => {
    if (event.key === 'Enter') {
      updateComments((event.target as HTMLInputElement).value);
    }
  };

  if (loading)
    return (
      <Box sx={{ display: 'flex' }}>
        <CircularProgress />
      </Box>
    );
  if (error)
    return (
      <Alert variant="outlined" severity="error">
        Could not load comments
      </Alert>
    );

  return (
    <div aria-label="Comments section">
      <Box
        sx={{
          display: 'flex',
          alignItems: 'center',
          justifyContent: 'flex-end',
          width: '80%',
          maxWidth: '100%',
        }}
      >
        <TextField
          fullWidth
          label="Review this movie"
          id="fullWidth"
          value={newComment}
          onChange={(e) => setNewComment(e.target.value)}
          onKeyDown={handleEnterPress}
        />
        <Button sx={{ marginLeft: '12px' }} variant="outlined" onClick={() => updateComments(newComment)}>
          Submit Review
        </Button>
      </Box>

      <div style={{ textAlign: 'center' }}>
        {allComments
          .slice()
          .reverse()
          .map(
            (commentObject, index) =>
              (showAllComments || index < 2) && (
                <Paper key={index} style={{ padding: '10px 20px', marginTop: 10, marginBottom: 30 }}>
                  <Grid container wrap="nowrap" spacing={2}>
                    <Grid item>
                      <Avatar sx={{ bgcolor: blue[500], marginTop: '0.5rem' }}>{commentObject.userID[0]}</Avatar>
                    </Grid>
                    <Grid tabIndex={0} item>
                      <p>
                        {commentObject.userID}: {commentObject.comment}
                      </p>
                    </Grid>
                  </Grid>
                </Paper>
              ),
          )}

        {allComments.length > 2 && (
          <Button
            variant="outlined"
            onClick={handleToggleComments}
            style={{ color: 'white', marginBottom: '1rem', marginTop: '1rem' }}
          >
            {showAllComments ? 'Show less comments' : 'Show more comments'}
          </Button>
        )}

        {allComments.length === 0 && (
          <Alert sx={{ marginTop: '1rem', marginBottom: '1rem' }} variant="outlined" severity="info" tabIndex={-1}>
            No reviews for this movie yet - be the first one to comment
          </Alert>
        )}
      </div>
    </div>
  );
};

export default CommentSection;
