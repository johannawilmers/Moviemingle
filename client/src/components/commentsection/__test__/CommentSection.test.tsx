import '@testing-library/jest-dom';

import { act, fireEvent, screen, waitFor } from '@testing-library/react';

import { ADD_COMMENT, GET_ALL_COMMENTS } from '../../../queries/Comments';
import { renderWithMocks, renderWithPageContent } from '../../../utils/testUtils';
import { movie } from '../../../utils/types';
import CommentSection from '../CommentSection';

const mockmovie: movie = {
  backdrop_path: '/4XM8DUTQb3lhLemJC51Jx4a2EuA.jpg',
  genre_ids: [28, 80, 53],
  _id: 385687,
  original_title: 'Fast X',
  overview:
    "Over many missions and against impossible odds, Dom Toretto and his family have outsmarted, out-nerved and outdriven every foe in their path. Now, they confront the most lethal opponent they've ever faced: A terrifying threat emerging from the shadows of the past who's fueled by blood revenge, and who is determined to shatter this family and destroy everything—and everyone—that Dom loves, forever.",
  poster_path: '/fiVW06jE7z9YnO4trhaMEdclSiC.jpg',
  release_date: '2023-05-17',
  title: 'Fast X',
  vote_average: 7.3,
  isFavorite: true,
};

const comments = [
  {
    movieId: 385687,
    userID: 'testUser',
    comment: ['Such a good action movie!'],
  },
];

const mocks = [
  {
    request: {
      query: GET_ALL_COMMENTS,
      variables: { movieId: 385687 },
    },
    result: {
      data: {
        getAllComments: comments,
      },
    },
  },
];

describe('CommentSection', () => {
  it('renders with initial mock comments', async () => {
    let container;
    await act(async () => {
      ({ container } = renderWithMocks(
        <CommentSection movieId={mockmovie._id || 0} userID={comments[0].userID} />,
        mocks,
      ));
    });
    await screen.findByText('testUser: Such a good action movie!');
    expect(screen.queryByText('Loading')).toBeNull();
    comments.forEach((comment) => {
      expect(screen.getByText(`${comment.userID}: ${comment.comment}`)).toBeInTheDocument();
    });
    expect(container).toMatchSnapshot();
  });

  it('render the commentsection before adding comment', async () => {
    await renderWithPageContent(<CommentSection movieId={mockmovie._id || 0} userID={comments[0].userID} />);
    waitFor(() => {
      expect(screen.getByText('Submit Review')).toBeInTheDocument();
      expect(screen.getByText('No reviews to this movie yet - be the first one to comment')).toBeInTheDocument();
    });
  });

  it('render the addComment button', () => {
    renderWithPageContent(<CommentSection movieId={mockmovie._id || 0} userID={comments[0].userID} />);
    expect(screen.getByText('Submit Review')).toBeInTheDocument();
  });

  it('render the textField placeholder', () => {
    renderWithPageContent(<CommentSection movieId={mockmovie._id || 0} userID={comments[0].userID} />);
    const inputField = screen.getByLabelText('Review this movie');
    expect(inputField).toBeInTheDocument();
  });

  it('render the commentsection with an empty submit', async () => {
    const jsdomAlert = window.alert;
    window.alert = () => {};
    renderWithMocks(<CommentSection movieId={mockmovie._id || 0} userID={comments[0].userID} />, mocks);
    await waitFor(() => {
      fireEvent.change(screen.getByLabelText('Review this movie'), { target: { value: '' } });
      fireEvent.click(screen.getByText('Submit Review'));
      expect(screen.getAllByRole('alert')).toHaveLength(1);
    });
    window.alert = jsdomAlert;
  });

  it('render the commentsection with a submit', async () => {
    const comments = [
      {
        movieId: 385687,
        userID: 'testUser',
        comment: ['Such a good action movie!'],
      },
      {
        movieId: 385687,
        userID: 'testUser',
        comment: ['Wow i really liked this'],
      },
    ];

    const mocks = [
      {
        request: {
          query: ADD_COMMENT,
          variables: {
            addCommentId: 385687,
            comment: 'Wow i really liked this',
            userID: 'testUser',
          },
        },
        result: {
          data: {
            addComment: {
              movieId: 385687,
              userID: 'testUser',
              comment: 'Wow i really liked this',
            },
          },
        },
      },
      {
        request: {
          query: GET_ALL_COMMENTS,
          variables: { movieId: 385687 },
        },
        result: {
          data: {
            getAllComments: comments,
            userID: 'testuser',
          },
        },
      },
    ];
    renderWithMocks(<CommentSection movieId={mockmovie._id || 0} userID={comments[0].userID} />, mocks);
    await waitFor(() => {
      fireEvent.change(screen.getByLabelText('Review this movie'), { target: { value: 'Wow i really liked this' } });
      fireEvent.click(screen.getByText('Submit Review'));
      waitFor(() => {
        expect(screen.getByText(comments[0].userID + ': ' + 'Wow i really liked this')).toBeInTheDocument();
      });
    });
  });

  it('test that "show more comments" changes to "show less comments', async () => {
    const comments = [
      {
        movieId: 385687,
        userID: 'testUser',
        comment: ['Such a good action movie!'],
      },
      {
        movieId: 385687,
        userID: 'testUser',
        comment: ['Wow i really liked this'],
      },
      {
        movieId: 385687,
        userID: 'testUser',
        comment: ['Liked this a lot'],
      },
      {
        movieId: 385687,
        userID: 'testUser',
        comment: ['Not so nice'],
      },
    ];
    const updatedMocks = [
      {
        request: {
          query: ADD_COMMENT,
          variables: {
            addCommentId: 385687,
            comment: 'shit',
            userID: 'testUser',
          },
        },
        result: {
          data: {
            addComment: {
              movieId: 385687,
              userID: 'testUser',
              comment: 'Not so nice',
            },
          },
        },
      },
      {
        request: {
          query: GET_ALL_COMMENTS,
          variables: { movieId: 385687 },
        },
        result: {
          data: {
            getAllComments: comments,
          },
        },
      },
    ];
    renderWithMocks(<CommentSection movieId={mockmovie._id || 0} userID={comments[0].userID} />, updatedMocks);
    await waitFor(() => {
      fireEvent.click(screen.getByText('Show more comments'));
      expect(screen.getByText('Show less comments')).toBeInTheDocument();
    });
  });
});
