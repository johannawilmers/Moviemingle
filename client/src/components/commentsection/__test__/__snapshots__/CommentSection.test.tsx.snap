// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CommentSection > renders with initial mock comments 1`] = `
<div>
  <div
    aria-label="Comments section"
  >
    <div
      class="MuiBox-root css-16fxm2j"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for="fullWidth"
          id="fullWidth-label"
        >
          Review this movie
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id="fullWidth"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-yjsfm1"
            >
              <span>
                Review this movie
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-1b165oe-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Submit Review
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
    <div
      style="text-align: center;"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1ps6pg7-MuiPaper-root"
        style="padding: 10px 20px; margin-top: 10px; margin-bottom: 30px;"
      >
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-wrap-xs-nowrap css-1cfjp91-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
          >
            <div
              class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-1nt9a3v-MuiAvatar-root"
            >
              t
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
            tabindex="0"
          >
            <p>
              testUser
              : 
              Such a good action movie!
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
